### PR TITLE
Fixes 210: added support for continous deployment from yaydoc

### LINKED ISSUE #210: Add support for continuous deployment from yaydoc
### This is a
- [ ] Bug report
- [x] Feature request

### Current behavior

right now we have been using Travis to push documentation 
### Expected behavior

use yaydoc instead of using Travis to push documentation 

### Steps to reproduce
1.
2.
3.

### PR DESCRIPTION
**Pull request by sch00lb0y**
## Description

Integrated GitHub webhook to support continuous documentation deployment from yaydoc
## Related Issue

close #210 
## Motivation and Context

right now we depending on third party to do continuous deployment. by having this feature which will enable to implement more feature as our wish in future and also no need for depending any third party ci.
## Screenshots (if appropriate):
ci

## Types of changes

- [ ] Bug fix 
- [X] New feature

## Checklist:

- [x] I have read the [Contribution & Best Practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] There is a corresponding issue for this pull request.
- [x] Mentioned the Issue number in the pull request commit message `Fixes #<number> commit message`
- [x] There is only one commit per issue.

### CODE REVIEW

**Comment by mariobehling** [comment]
Please also resolve conflicts.

**Comment by sch00lb0y** [comment]
@pri22296 @imujjwal96 please check this demo deployment http://52.221.209.153/ and I'll resolve conflicts once you reviewers okay with this PR 
Thanks.

**Comment by pri22296** [comment]
@sch00lb0y Please consider these points

- Remove permission to delete my repository
- Make the CI deploy button visible by default.

**Comment by sch00lb0y** [comment]
@pri22296 now I added one more route for handling ci and I added `'write:repo_hook'` scope for permission. This is mandatory for to write the webhook. If it asks to delete repository permission we can't do anything for that. and CI deploy button is visible by default only. If this answer not solving your second query. please elaborate

**Comment by pri22296** [comment]
@sch00lb0y Sorry about my second query. Didn't realize that the button was there. My mistake.

**Comment by sch00lb0y** [comment]
github
@pri22296  It is not requesting any delete repo request. 
@imujjwal96 please review this PR.

**Comment by pri22296** [comment]
I tried it with my repo. Docs were not updated on push

**Comment by sch00lb0y** [comment]
I have cleared the DB while checking that's why didn't work I guess. just now I tested with my test repo it is working fine.Please delete the webhook on your repository and register once again with yaydoc. it'll work for sure

**Comment by pri22296** [comment]
@sch00lb0y Checked it. worked this time

File: routes/ci.js [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        {
+          githubId: req.body.repository.owner.id,
+          name: req.body.repository.name
+        }
+      ).
+      then(function(result) {
+        var data = {
+          email: result.email,
+          gitUrl: req.body.repository.clone_url,
+          docTheme: "alabaster",
```
Comment (pri22296) [inline routes/ci.js]: Don't hard code `alabaster`. Instead set it to be an empty string. `generate.sh` will handle the rest

File: routes/ci.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            next({
+              status: 500,
+              message: 'something went wrong with db'
+            })
+          })
+      })
+    } else {
+      res.render("index", {
+        showMessage: true,
+        messages: `Yaydoc was already integrate with this repository`
```
Comment (pri22296) [inline routes/ci.js]: Yaydoc is already integrated with this repository

File: routes/ci.js [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+      }, function(error, response, body) {
+        repositoryModel.newRepository(repositoryName,
+          req.session.username,
+          req.session.githubId,
+          crypter.encrypt(req.session.token),
+          req.session.email)
+          .then(function(result) {
+            res.render("index", {
+              showMessage: true,
+              messages: `Thanks for registering with Yaydoc.Here after Documentation will be pushed on each commit`
```
Comment (pri22296) [inline routes/ci.js]: documentation will be pushed where? specify that.

File: routes/ci.js [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            })
+          })
+          .catch(function(err) {
+            next({
+              status: 500,
+              message: 'something went wrong with db'
```
Comment (pri22296) [inline routes/ci.js]: Something went wrong

File: routes/ci.js (hunk 1 above)
Comment (sch00lb0y) [inline routes/ci.js]: Made changes

File: routes/ci.js (hunk 2 above)
Comment (sch00lb0y) [inline routes/ci.js]: Made changes

File: routes/ci.js (hunk 3 above)
Comment (sch00lb0y) [inline routes/ci.js]: Made changes

File: routes/ci.js (hunk 4 above)
Comment (sch00lb0y) [inline routes/ci.js]: Made changes

File: backend/deploy.js [hunk 5]
```diff
@@ -1,6 +1,6 @@
 var crypter = require("../util/crypter.js")
 var spawn = require('child_process').spawn;
-
+var socketHandler = require('./socketHandler.js')
```
Comment (imujjwal96) [inline backend/deploy.js]: why is this required?

File: model/repository.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+})
+
+const model = mongoose.model('repository', repositorySchema);
+
+exports.newRepository = function (name, username, githubId, accessToken, email) {
+  return new Promise(function (resolve, reject) {
+    new model({
+      name: name,
+      username: username,
+      githubId: githubId,
```
Comment (imujjwal96) [inline model/repository.js]: what's githubId?

File: routes/ci.js [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        showMessage: true,
+        messages: `Yaydoc is already integrated with this repository.`
+      })
+    }
+  })
+})
+
+router.post('/webhook', function(req, res, next) {
+  var event = req.get('X-GitHub-Event')
+  switch (event) {
```
Comment (imujjwal96) [inline routes/ci.js]: why use switch case here?

File: backend/deploy.js (hunk 5 above)
Comment (sch00lb0y) [inline backend/deploy.js]: now in the first parameter, you have to pass a socket object. I don't want to create as a separate function to handle generation. so I made a module named socket handler. it'll handle if the first parameter is socket otherwise it'll just ignore the object do nothing. we'll get a modular code

File: model/repository.js (hunk 6 above)
Comment (sch00lb0y) [inline model/repository.js]: username can be changed so we can't use as a unique id. GitHub having uniqueId for each user. In order to sync with users, I'm using GitHub ID as GitHubId

File: routes/ci.js (hunk 7 above)
Comment (sch00lb0y) [inline routes/ci.js]: now, we use only push event. In future, we may use more webhook events so Instead of using if case I used switch

File: backend/deploy.js (hunk 5 above)
Comment (imujjwal96) [inline backend/deploy.js]: When will it be undefined?

Comment (sch00lb0y) [inline backend/deploy.js]: check webhook handler. I'm just passing empty object
```js
            deploy.deployPages({}, {
              email: result.email,
              gitURL: req.body.repository.clone_url,
              username: result.username,
              uniqueId: generatedData.uniqueId,
              encryptedToken: result.accessToken
            })
```

**Comment by imujjwal96** [comment]
instead of showing the message in a modal, show it in as a notification

**Comment by imujjwal96** [comment]
the flow is very vague. What i assumed (and certain that most users will assume) is that we have to fill the form before clicking on CI deploy. Think of a different flow

**Comment by imujjwal96** [comment]
What will happens if i have more than 100 repositories in my github profile

**Comment by sch00lb0y** [comment]
>What will happens if i have more than 100 repositories in my github profile

since it is a single issue, making all feature in one PR is bit difficult after the PR merge I'll add search feature like Heroku.

> instead of showing the message in a modal, show it in as a notification

if any event triggers we can use notification, since we showing Thank you Message and Already registered Message I think Modal will be appropriate if we use notifiication it'll be unnoticed. 

> the flow is very vague. What i assumed (and certain that most users will assume) is that we have to fill the form before clicking on CI deploy. Think of a different flow

ci
Now I have placed the button like this.is it okay ?

**Comment by imujjwal96** [comment]
The reason notifications exist is so that they are noticed.

**Comment by imujjwal96** [comment]
Do not headlining the form as WEBUI. The web UI consist of everything a user sees own our website, including the CI Deploy

**Comment by sch00lb0y** [comment]
@imujjwal96 I removed modal and changed to the notification. I changed the headline to Web Generator. If you any other name in mind Please let me know

**Comment by imujjwal96** [comment]
Please resolve conflicts. I'm happy otherwise.

**Comment by sch00lb0y** [comment]
@pri22296 @imujjwal96 I have resolved conflicts

File: backend/deploy.js [hunk 8]
```diff
@@ -19,15 +20,15 @@ exports.deployPages = function (socket, data) {
 
   var process = spawn("./ghpages_deploy.sh", args);
 
-  output.lineOutput(socket, process, 'deploy-logs', 18);
-  output.lineError(socket, process, 'err-logs');
+  socketHandler.handleLineOutput(socket, process, 'github-deploy-logs', 18);
```
Comment (pri22296) [inline backend/deploy.js]: Consider moving the logic of `socketHandler` to util/output.js

File: backend/generator.js [hunk 9]
```diff
@@ -4,11 +4,11 @@ var mailer = require('./mailer');
 var uuidV4 = require("uuid/v4");
 var validation = require("../public/scripts/validation.js");
 var spawn = require('child_process').spawn;
-var output = require('../util/output');
+var socketHandler = require('./socketHandler.js');
 
-exports.executeScript = function (socket, formData) {
+exports.executeScript = function (socket, formData, cb) {
```
Comment (pri22296) [inline backend/generator.js]: I suppose `cb` stands for `callback`? If that is the case, consider naming it `callback`. shouldn't be a lot of extra keystrokes.

Comment (sch00lb0y) [inline backend/generator.js]: just a js standard where people widely use. https://stackoverflow.com/questions/31780872/what-is-cb-in-node Anyways I made changes

File: backend/deploy.js (hunk 8 above)
Comment (sch00lb0y) [inline backend/deploy.js]: I moved `socketHandler.js` to util folder. but moving logic to `output.js` won't make much sense.

Comment (pri22296) [inline backend/deploy.js]: They seem to be doing similar things that's why I asked to move it If it was possible. Anyways seems fine.

**Comment by pri22296** [comment]
@sch00lb0y Please merge your changes from upstream

**Comment by pri22296** [comment]
@sch00lb0y Just a tip. In future, use `git pull --rebase upstream master` to avoid creating unnecessary merge commits.

**Comment by imujjwal96** [comment]
@sch00lb0y Can you show us a Heroku deployment with this change?

**Comment by sch00lb0y** [comment]
@imujjwal96 please check this http://demo08.herokuapp.com/

**Comment by imujjwal96** [comment]
image
**S**elect the repository
**R**egister

**Comment by imujjwal96** [comment]
why's the route /ci/register
image

**Comment by sch00lb0y** [comment]
`why's the route /ci/register` POST request for form submission. but then I can use '/'  but it won't look good and also it'll become difficult to understand the code

**Comment by imujjwal96** [comment]
Why do I perform a basic documentation generation at the /ci/register route? This doesn't look good and frankly doesn't make sense.

**Comment by sch00lb0y** [comment]
@imujjwal96 I'll change the route to POST `/`

**Comment by imujjwal96** [comment]
Oh please don't do this. What I'm saying is that there is a particular design pattern that web application follows. 
We first send a **P**OST request, perform an action, **R**edirect and then render a template in a **G**ET request.
https://en.wikipedia.org/wiki/Post/Redirect/Get

Want you've done is that you're rendering the index page in the POST request itself. What this will lead to is that every time the user refreshes that page, the POST request is sent again. This is something that is never desired.


What I want you to do is that instead of rendering the index page in you POST request route, redirect to the `/` route with a registration successful notification

**Comment by sch00lb0y** [comment]
@imujjwal96 Yeah I already made changes according to that I'll test and let you know soon

File: routes/ci.js [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+router.get('/register', function (req, res, next) {
+  request({
+    url: `https://api.github.com/users/${req.session.username}/repos?per_page=100`,
+    headers: {
+      'User-Agent': 'request'
+    }
+  },
+  function (error, response, body) {
+    var repos = JSON.parse(body);
+    res.render("index", {
```
Comment (imujjwal96) [inline routes/ci.js]: I think we should keep the index template specific for the `/` route. What is happening here is that since the modal is collapsible, i can perform the actions which are aesthetically supposed to be performed in the `/` route, at the `/ci/register` route.

Comment (sch00lb0y) [inline routes/ci.js]: Can I do the same `res.redirect("/?message=Thanks for registering with Yaydoc.Hereafter Documentation will be pushed to the GitHub pages on each commit.");` but the logic for index route will become big. if you have any suggestion please do give me

Comment (imujjwal96) [inline routes/ci.js]: do res.redirect("/?success=true") and do something similar to what i did at ownCloud
https://github.com/owncloud/core/pull/27277/files#diff-5af29c9f0fdb4338457f3bb7024dc991R152

Comment (imujjwal96) [inline routes/ci.js]: That is, remove the query params as soon as the server reads it and shows notification

Comment (sch00lb0y) [inline routes/ci.js]: @imujjwal96 I passing the repo array in query string and changing url state . please check https://demo08.herokuapp.com

File: views/index.jade [hunk 11]
```diff
@@ -86,5 +96,33 @@ block content
...
+             each repo in repos
+               option #{repo} 
+           .modal-footer  
+             button.btn.btn-default(type="submit") Register  
+  if repos !== undefined
+   script.
+    $("#ciModal").modal('show');
+  if showMessage === true
+   script.
+     $('.notification').append($('<li>')).text("#{message}");
```
Comment (imujjwal96) [inline views/index.jade]: use `styles.showNotification()` function defined in `public/scripts/styles.js`

File: routes/ci.js [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+          }
+        }
+      }, function(error, response, body) {
+        repositoryModel.newRepository(repositoryName,
+          req.session.username,
+          req.session.githubId,
+          crypter.encrypt(req.session.token),
+          req.session.email)
+          .then(function(result) {
+            res.redirect("/?message=Thanks for registering with Yaydoc.Hereafter Documentation will be pushed to the GitHub pages on each commit.");
```
Comment (imujjwal96) [inline routes/ci.js]: instead of spitting out the message as a get request, use `/?status=true` or something like '/?status=200' 
and in `index.jade` line 119 check for this status and display the message

File: views/index.jade (hunk 11 above)
Comment (sch00lb0y) [inline views/index.jade]: made changes

File: routes/ci.js (hunk 12 above)
Comment (sch00lb0y) [inline routes/ci.js]: made changes

**Comment by sch00lb0y** [comment]
@imujjwal96 I have updated the branch.Please check

**Comment by mariobehling** [comment]
http://52.221.209.153/ links to localhost when I click on CI.

**Comment by sch00lb0y** [comment]
@mariobehling I have updated the README.md . Please check the readme [here](https://github.com/sch00lb0y/yaydoc/tree/continuous). I was testing in localhost that's why it is redirected to localhost. I have changed that also. check the deployment here https://demo08.herokuapp.com/ . 

Since it it is a big PR not able to add this feature 
>  Add info how user can choose theme and change location/branch of documentation site

I'll do this feature after this pr get merged

**Comment by mariobehling** [comment]
Ok, thanks. I guess theme changes etc. can be taken care of in following issues. Cheers.
